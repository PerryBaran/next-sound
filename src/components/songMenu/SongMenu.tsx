"use client"

import { useState } from "react"
import css from "./songMenu.module.css"
import { usePlaylistContext } from "../../context/PlaylistContext"
import { play, addList, dots } from "../../media/icons/index"
import Image from "next/image"

interface Props {
  songs: {
    songName: string
    audio: string
    image?: string
    artistName: string
    albumName: string
  }[]
}

export default function SongMenu({ songs }: Props) {
  const { handleAddToPlaylist } = usePlaylistContext()
  const [visible, setVisible] = useState(false)

  const handleBlur = () => {
    setVisible(false)
  }

  return (
    <div className={css.container} onBlur={handleBlur}>
      {visible ? (
        <div className={css.options}>
          <button onMouseDown={() => handleAddToPlaylist(songs, true, true)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={play} alt="play now" width={9} />
            </span>
            <span>Play Now</span>
          </button>
          <button
            onMouseDown={() => handleAddToPlaylist(songs, true, false)}
            className={css["play-next"]}
          >
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Play Next</span>
          </button>
          <button onMouseDown={() => handleAddToPlaylist(songs, false, false)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Add To Queue</span>
          </button>
        </div>
      ) : null}
      <button
        onClick={() => setVisible((prev) => !prev)}
        className={css.button}
      >
        <Image src={dots} alt="..." height={18} width={18} />
      </button>
    </div>
  )
}
