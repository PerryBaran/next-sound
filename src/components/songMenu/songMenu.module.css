.container {
  position: relative;
}

.options {
  position: absolute;
  bottom: 20px;
  right: 0px;
  width: 110px;
  display: flex;
  flex-direction: column;
  border: 1px solid var(--darker-grey);
  overflow: hidden;
  z-index: 1000;
}

.options > button {
  border: none;
  border-bottom: 1px solid var(--darker-grey);
  background-color: var(--almost-white);
  color: var(--darker-grey);
  font-weight: 100;
  cursor: pointer;
  text-align: left;
  display: flex;
  height: 20px;
}

.options > button:hover {
  background-color: var(--main-colour);
  color: var(--almost-white);
}

.options > button:last-child {
  border-bottom: none;
}

.image-container {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 100%;
}

.options > button:hover img {
  filter: invert();
}

.play-next img {
  transform: scaleY(-1);
}

.button {
  border: none;
  background: none;
  padding: 5px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.button > img {
  transition-duration: 250ms;
  transform: rotate(90deg);
}

.button:hover > img {
  transform: rotate(90deg) scale(1.25);
}
