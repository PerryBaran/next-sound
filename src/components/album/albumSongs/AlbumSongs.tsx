import css from "./album-songs.module.css"
import SongMenu from "../../songMenu/SongMenu"

interface Props {
  songName: string
  audio: string
  image?: string
  artistName: string
  albumName: string
}

export default function AlbumSongs(props: Props) {
  return (
    <li className={css["container"]}>
      <h4>{props.songName}</h4>
      <SongMenu songs={[props]} />
    </li>
  )
}
