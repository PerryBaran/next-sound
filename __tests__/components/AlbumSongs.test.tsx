import { render, screen } from "@testing-library/react"
import AlbumSongs from "../../src/components/album/albumSongs/AlbumSongs"
import * as SongMenu from "../../src/components/songMenu/SongMenu"

describe("AlbumSongs", () => {
  const props = {
    songName: "song-name",
    audio: "song-audio",
    image: "album-image",
    artistName: "artist-name",
    albumName: "albumName"
  }

  const mockedSongMenu = jest.fn()

  jest.spyOn(SongMenu, "default").mockImplementation((props: any) => {
    mockedSongMenu(props)
    return <div />
  })

  test("renders correctly", () => {
    render(<AlbumSongs {...props} />)

    expect(screen.getByRole("listitem")).toBeInTheDocument()
    expect(screen.getByRole("heading")).toHaveTextContent(props.songName)
    expect(mockedSongMenu).toHaveBeenCalledWith({ songs: [props] })
  })

  test("snapshot", () => {
    const { asFragment } = render(<AlbumSongs {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
