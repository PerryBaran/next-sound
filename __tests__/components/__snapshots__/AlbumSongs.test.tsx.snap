// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumSongs snapshot 1`] = `
<DocumentFragment>
  <li
    class="container"
  >
    <h4>
      song-name
    </h4>
    <div />
  </li>
</DocumentFragment>
`;
